rails_programming/mailers_advanced_topics/websockets_and_actioncable.md

```diff
@@ -0,0 +1,315 @@
+### Introduction
+
+When you think about a typical interaction between a client and website, a browser makes a request to the server and after that request has completed the server closes the connection.
+
+There are often times where you want to keep that connection open so that the server can update the client if there is something relevant for the user. For that we have Websockets and ActionCable brings Websockets to Rails in an easy to use way.
+
+### Learning Outcomes
+Look through these now and then use them to test yourself after doing the assignment.
+
+* What is a websocket?
+* What options did developers have before websockets to update a client without a user request?
+* What is a consumer?
+* What is a subscriber?
+* What is a channel?
+* What is a stream?
+* How can you broadcast to a stream from the server?
+
+### What is a Websocket?
+
+Before diving into Actioncable specifics we should first understand what a websocket is and what problem it seeks to solve.
+
+The Websocket API was created to allow two-way communication between a server and a client. A typically http connection between a server and client opens and closes as the client requests something. A websocket connection keeps the http connection alive so that the server can send updates to the client without a specific request being made.
+
+A websocket manages the connection between server and client as well as sending and receiving data through that connection.
+
+To understand what problem websockets solve let us consider an example of where websockets may be useful and then consider what our options would have been prior to the Websocket API.
+
+Let us imagine we've built an awesome new social media app. Chewbooka. A place to be our true selves and talk all things Chewbacca. One key feature for our app, to make us all billions, is that when someone creates a new post extolling the virtues of Peter Mayhew, that all the other users connected to our app receive that post without having to manually refresh for updates.
+
+We'll be rich! But, in a time before webosckets, how are we going to solve it.
+
+We could use Javascript to set an interval to reach out to the server at regular intervals to see if there are any updates. If there are we could notify users of the new post and update their view and if there wasn't we can just return an empty response. This still involves opening and closing a request on the sever. This was a technique know as polling and was one of the first ways websites tried to bring server updates to the client. The downside to this was one of efficiency. If there were no updates for the client it would still request an update from the server. In an age of capped internet usage this was a big no.
```

Comment:
Explaining how it was done before is a really nice touch, good low key way of exposing them to polling 💯